subt/tools/analyze_detectors.py

```diff
@@ -0,0 +1,243 @@
+"""evaluate detectors"""
+import os
+import numpy as np
+import cv2
+from matplotlib import pyplot as plt
+
+try:
+    import torch
+    import subt.artf_model
+    from subt.artf_detector import Detector
+except ImportError:
+    print('\nWarning: missing torch!\n')
+
+from osgar.logger import LogReader, lookup_stream_id
+from osgar.lib.serialize import deserialize
+from subt.tf_detector import CvDetector
+from subt.artf_node import check_results
+
+g_streams = ["logimage.image", "logimage_rear.image"]
+g_artf_names = ["survivor","backpack", "phone", "helmet", "rope", "fire_extinguisher", "drill", "vent", "cube"]
+
+NAMES_AND_SCORES = {'backpack': 0.1,
+                    'survivor': 0.1,
+                    'phone': 0.1,
+                    'rope': 0.1,
+                    'helmet': 0.1,
+                    'fire_extinguisher': 0.1,
+                    'drill': 0.1,
+                    'vent': 0.1,
+                    'cube': 0.1,
+                    'robot': 1.0,
+                    'breadcrumb': 1.0,
+                    'nothing': 1.0
+                    }
+MIN_SCORES = np.array(list(NAMES_AND_SCORES.values()))
+
+
+def take_third(item):
+    return item[2]
+
+
+def proces_torch_result(result):
+    result.sort(key=take_third, reverse=True)
+    score_torch = result[1][2]
+    points = [[x, y] for x, y, c in result][:2]
+
+    return score_torch, points
+
+
+def manual_sorting(data):
+    ii = 0
+    while True:
+        if ii < 0:
+            ii = 0
+        if ii >= len(data):
+            ii = len(data) - 1
+        artf_name, score_t, score_cv, points, bbox, im_path, detection_type = data[ii]
+        print(bbox)
+        x, y, xw, yh = bbox
+        im_path = im_path[:-1]  # remove line end
+        img = cv2.imread(str(im_path), 1)
+        assert img is not None
+        cv2.rectangle(img, (x, y), (xw, yh), (0, 0, 255))
+        cv2.putText(img, artf_name, (x, y), fontFace=cv2.FONT_HERSHEY_SIMPLEX, fontScale=1, color=(0, 0, 255))
+        for xc, yc in points:
+            cv2.circle(img, (xc, yc), radius=4, color=(255, 255, 0), thickness=1)
+        cv2.putText(img, str(detection_type), (10, 50), fontFace=cv2.FONT_HERSHEY_SIMPLEX, fontScale=1, color=(255, 0, 0))
+        im_to_show = cv2.resize(img, (1280, 720))
+        cv2.imshow("win", im_to_show)
+
+        k = cv2.waitKey(0) & 0xFF
+        if k == ord("n"):  # next img
+            ii += 1
+        elif k == ord("b"):  # back one img
+            ii -= 1
+        elif k == ord("t"):
+            data[ii][6] = "true"  # true detection
+        elif k == ord("f"):
+            data[ii][6] = "false"  # false detection
+        elif k == ord("d"):
+            data[ii][3] = "None"  # do not use
+        elif k == ord("q"):  # close and save
+            break
+
+    cv2.destroyAllWindows()
+
+    return [[artf_name, score_t, score_cv, detection_type] for artf_name, score_t, score_cv, __, __, __, detection_type in data]
+
+
+def create_detector():
```

Comment:
Is there a way to share this function with subt/artf_whatever.py? It is very unlikely that we would keep this version here up to date.

Reply:
My effort was to minimize the changes outside of the analyze-detectors.. However, we could  put the function outside of the class and add an optional parameter for "confidence_thresholds". I will show my proposal in a separate commit.